Source/Core/DolphinQt/FIFO/FIFOAnalyzer.cpp

```diff
@@ -160,168 +165,256 @@ void FIFOAnalyzer::UpdateTree()
-        ASSERT(false);
-        objectdata += 8;
-        new_label = QStringLiteral("CALL DL");
-        break;
+    case OpcodeDecoder::GX_LOAD_BP_REG:
+    {
+      const u8 cmd2 = object[object_offset++];
+      const u32 cmddata = Common::swap24(&object[object_offset]);
+      object_offset += 3;
+
+      const auto [name, desc] = GetBPRegInfo(cmd2, cmddata);
+      ASSERT(!name.empty());
 
-      case OpcodeDecoder::GX_LOAD_BP_REG:
+      new_label = QStringLiteral("BP  %1  %2  %3")
+                      .arg(cmd2, 2, 16, QLatin1Char('0'))
+                      .arg(cmddata, 6, 16, QLatin1Char('0'))
+                      .arg(QString::fromStdString(name));
+    }
+    break;
+
+    default:
+      if ((command & 0xC0) == 0x80)
       {
-        const u8 cmd2 = *objectdata++;
-        const u32 cmddata = Common::swap24(objectdata);
-        objectdata += 3;
+        // Object primitive data
 
-        const auto [name, desc] = GetBPRegInfo(cmd2, cmddata);
-        ASSERT(!name.empty());
+        const u8 vat = command & OpcodeDecoder::GX_VAT_MASK;
+        const auto& vtx_desc = frame_info.objectCPStates[object_nr].vtxDesc;
+        const auto& vtx_attr = frame_info.objectCPStates[object_nr].vtxAttr[vat];
 
-        new_label = QStringLiteral("BP  %1  %2  %3")
-                        .arg(cmd2, 2, 16, QLatin1Char('0'))
-                        .arg(cmddata, 6, 16, QLatin1Char('0'))
-                        .arg(QString::fromStdString(name));
-      }
-      break;
+        const auto name = GetPrimitiveName(command);
 
-      default:
-        new_label = tr("Unexpected 0x80 call? Aborting...");
-        objectdata = static_cast<const u8*>(next_objdata_start);
-        break;
+        const u16 vertex_count = Common::swap16(&object[object_offset]);
+        object_offset += 2;
+        const u32 vertex_size = VertexLoaderBase::GetVertexSize(vtx_desc, vtx_attr);
+
+        // Note that vertex_count is allowed to be 0, with no special treatment
+        // (another command just comes right after the current command, with no vertices in between)
+        const u32 object_prim_size = vertex_count * vertex_size;
+
+        new_label = QStringLiteral("PRIMITIVE %1 (%2)  %3 vertices %4 bytes/vertex %5 total bytes")
+                        .arg(QString::fromStdString(name))
+                        .arg(command, 2, 16, QLatin1Char('0'))
+                        .arg(vertex_count)
+                        .arg(vertex_size)
+                        .arg(object_prim_size);
+
+        // It's not really useful to have a massive unreadable hex string for the object primitives.
+        // Put it in the description instead.
+
+// #define INCLUDE_HEX_IN_PRIMITIVES
+#ifdef INCLUDE_HEX_IN_PRIMITIVES
+        new_label += QStringLiteral("   ");
+        for (u32 i = 0; i < object_prim_size; i++)
+        {
+          u32 byte = object[object_offset++];
+          new_label += QStringLiteral("%1").arg(byte, 2, 16, QLatin1Char('0'));
```

Comment:
Can be `const`, but it also doesn't need to be a `u32` here (`u8` would be better, or just not having a variable for it).